gateway/mw_graphql_complexity.go

```diff
@@ -0,0 +1,109 @@
+package gateway
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/graphql"
+
+	"github.com/TykTechnologies/tyk/user"
+)
+
+type ComplexityFailReason int
+
+const (
+	ComplexityFailReasonNone ComplexityFailReason = iota
+	ComplexityFailReasonInternalError
+	ComplexityFailReasonDepthLimitExceeded
+)
+
+type GraphQLComplexityMiddleware struct {
+	BaseMiddleware
+}
+
+func (m *GraphQLComplexityMiddleware) Name() string {
+	return "GraphQLComplexityMiddleware"
+}
+
+func (m *GraphQLComplexityMiddleware) EnabledForSpec() bool {
+	return m.Spec.GraphQL.Enabled && !m.Spec.UseKeylessAccess
+}
+
+// ProcessRequest will run any checks on the request on the way through the system, return an error to have the chain fail
+func (m *GraphQLComplexityMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Request, _ interface{}) (error, int) {
+	accessDef, _, err := GetAccessDefinitionByAPIIDOrSession(ctxGetSession(r), m.Spec)
+	if err != nil {
+		m.Logger().Debugf("Error while calculating GraphQL complexity: '%s'", err)
+		return m.handleComplexityFailReason(ComplexityFailReasonInternalError)
+	}
+
+	gqlRequest := ctxGetGraphQLRequest(r)
+
+	// If MaxQueryDepth is -1 or 0, it means unlimited and no need for depth limiting.
+	if m.DepthLimitEnabled(accessDef) {
+		if failReason := m.DepthLimitExceeded(gqlRequest, accessDef, m.Spec.GraphQLExecutor.Schema); failReason != ComplexityFailReasonNone {
+			return m.handleComplexityFailReason(failReason)
+		}
+	}
+
+	return nil, http.StatusOK
+}
+
+func (m *GraphQLComplexityMiddleware) DepthLimitEnabled(accessDef *user.AccessDefinition) bool {
+	// There is a possibility that depth limit is disabled on field level too,
+	// but we continue with this because of the explanation above.
```

Comment:
Seems this comment could be updated or removed

for example
`but we could not determine this without analyzing actual requested fields`

Reply:
yeah, right, thats why I hate comments :D